### PR TITLE
Add Shell and test

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

A DomainCreator for the Shell, which is made of six wedges.
Allows for equiangular or equidistant coordinates.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."

### CODE REVIEW

File: src/Domain/DomainCreators/Shell.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+#pragma once
+
+#include "Domain/DomainCreators/DomainCreator.hpp"
+#include "Options/Options.hpp"
+
+namespace DomainCreators {
+
+/// \ingroup DomainCreatorsGroup
+/// Create a 3D Domain consisting of six wedges and a central cube.
```
Comment (fmahebert) [inline src/Domain/DomainCreators/Shell.hpp]: Fix this

File: src/Domain/DomainCreators/Shell.hpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+class Shell : public DomainCreator<3, Frame::Inertial> {
+ public:
+  struct InnerRadius {
+    using type = double;
+    static constexpr OptionString_t help = {
+        "Radius of the sphere circumscribing the inner cube."};
```
Comment (fmahebert) [inline src/Domain/DomainCreators/Shell.hpp]: No cube

File: src/Domain/DomainCreators/Shell.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  struct InitialGridPoints {
+    using type = std::array<size_t, 2>;
+    static constexpr OptionString_t help = {
+        "Initial number of grid points in [r,angular]."};
+  };
+
+  struct WithEquiangularMap {
+    using type = bool;
+    static constexpr OptionString_t help = {
+        "Whether to use equiangular or equidistant variables."};
```
Comment (fmahebert) [inline src/Domain/DomainCreators/Shell.hpp]: variables -> coordinates

**Review by fmahebert — CHANGES_REQUESTED** [review]

File: tests/Unit/Domain/Test_DomainCreators.cpp [hunk 4]
```diff
@@ -224,3 +226,25 @@ SPECTRE_TEST_CASE("Unit.Domain.DomainCreators.TestHelperFunctions",
   tnsr::I<double, 3, Frame::Logical> expected{{{-1.0, -1.0, 1.0}}};
   CHECK(point_in_neighbor_frame<3>(custom_orientation, test_point) == expected);
 }
+
+SPECTRE_TEST_CASE("Unit.Domain.DomainCreators.Shell.Separation",
```
Comment (osheamonn) [inline tests/Unit/Domain/Test_DomainCreators.cpp]: Test that the neighbors, boundaries and maps are as expected.

**Review by kidder — CHANGES_REQUESTED** [review]
rebase, see comments on test of Disk and Sphere,

File: src/Domain/DomainCreators/Shell.hpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  struct InitialGridPoints {
+    using type = std::array<size_t, 2>;
+    static constexpr OptionString help = {
+        "Initial number of grid points in [r,angular]."};
+  };
+
+  struct UseEquiangularMap {
+    using type = bool;
+    static constexpr OptionString help = {
+        "Whether to use equiangular or equidistant coordinates."};
```
Comment (kidder) [inline src/Domain/DomainCreators/Shell.hpp]: oops

File: src/Domain/DomainCreators/Shell.hpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    static type default_value() { return true; }
+  };
+  using options = tmpl::list<InnerRadius, OuterRadius, InitialRefinement,
+                             InitialGridPoints, UseEquiangularMap>;
+
+  static constexpr OptionString help{
+      "Creates a 3D Shell with 6 Blocks. `use_equiangular_map_` has a default\n"
```
Comment (kidder) [inline src/Domain/DomainCreators/Shell.hpp]: help text talks about public options, not private member data

File: src/Domain/DomainCreators/Shell.hpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <vector>
+
+#include "DataStructures/Tensor/IndexType.hpp"
```
Comment (kidder) [inline src/Domain/DomainCreators/Shell.hpp]: move to cpp

File: tests/Unit/Domain/DomainCreators/Test_Shell.cpp [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+      "    InitialGridPoints: [2,3]\n"
+      "    UseEquiangularMap: true\n");
+  const double inner_radius = 1.0, outer_radius = 3.0;
+  const size_t refinement_level = 2;
+  const std::array<size_t, 2> grid_points_r_angular{{2, 3}};
+  test_shell_construction(
+      dynamic_cast<const DomainCreators::Shell<Frame::Inertial>&>(*shell),
+      inner_radius, outer_radius, true, grid_points_r_angular,
+      {6, make_array<3>(refinement_level)});
+}
```
Comment (kidder) [inline tests/Unit/Domain/DomainCreators/Test_Shell.cpp:133]: add equidistant tests

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by nilsdeppe — DISMISSED** [review]

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**